plugins/org.eclipse.emf.mwe2.language/src/org/eclipse/emf/mwe2/language/validation/Mwe2Validator.java

```diff
@@ -399,4 +401,31 @@ public boolean isMandatory(JvmIdentifiableElement feature) {
 		return ePackages;
 	}
 
+	@Check
+	public void checkDeprecatedElements(Component component) {
+		JvmType type = component.getType();
+		if (type instanceof JvmAnnotationTarget && isDeprecated((JvmAnnotationTarget) type)) {
+			warning(
+				"The type '" + type.getQualifiedName() + "' is deprecated.",
+				component,
+				Mwe2Package.Literals.REFERRABLE__NAME,
+				DEPRECATED_ELEMENT);
+		}
+	}
+
+	private boolean isDeprecated(JvmAnnotationTarget annotatable) {
```

Comment:
i wonder if this logic can be single-sourced between here and ui

Reply:
Yes, we can refactor that in a follow up PR. I think it could be worth to add the logic to the Xtext framework itself.

Reply:
wont help here as we bootstrap agains xtext 2.15